Slide animation for menus

onurtnrkl opened the pull request:
I added slide animation system for ui menus. Because some menus looks like slider(left menu, right menu etc) but not animated, seems annoying.

kd7uiy commented:
Thank you for the PR. It looks like it is failing Stylecop, and can't be merged. No worries, I do this pretty frequently myself, but... Take a look at https://github.com/OrderOfThePorcupine/ProjectPorcupine/blob/master/CONTRIBUTING.md#style-guidelines for how to install Stylecop and run it on your local machine, and commit another file to the list, which should help.

onurtnrkl commented:
> Thank you for the PR. It looks like it is failing Stylecop, and can't be merged. No worries, I do this pretty frequently myself, but... Take a look at https://github.com/OrderOfThePorcupine/ProjectPorcupine/blob/master/CONTRIBUTING.md#style-guidelines for how to install Stylecop and run it on your local machine, and commit another file to the list, which should help.

Thank you, I installed Stylecop and fixed. Also added license header. I will push changes soon.

kd7uiy commented:
Okay, I did some checks of it. First of all, it seems way too slow. I don't think people are going to want to wait 0.5s just to see a neat animation. Maybe speed it up to under 0.2 seconds, so it doesn't really affect anything. There are probably a few other things that we should consider as well, although I'll have to think on it some.

onurtnrkl commented:
> Okay, I did some checks of it. First of all, it seems way too slow. I don't think people are going to want to wait 1 second just to see a neat animation. Maybe speed it up to under 0.2 seconds, so it doesn't really affect anything. There are probably a few other things that we should consider as well, although I'll have to think on it some.
> 
> FYI, here is the feature. (Sorry for the gif cam in the wrong spot, but it worked well enough it wasn't worth messing with...)
> ![gif](https://user-images.githubusercontent.com/5951269/52800909-c60d8a00-304a-11e9-8c53-7a53f749bab6.gif)

Duration can be set from prefab that the script(SlideAnimation) is attached. I set it as 3 but you can make it one if you prefer it to slide faster.
<img width="264" alt="screenshot 2019-02-14 at 18 28 51" src="https://user-images.githubusercontent.com/8862032/52805228-6df47900-3086-11e9-9dd9-10f7de09b6f9.png">

koosemose commented:
Ok, a few issues here. First, animation speed should be set in code rather than prefab.

Second, it's generally not a good idea to include fixes with feature PRs, unless that fix is integral for the feature to function at all, it means we can't accept the fix until the feature is accepted.

Third (and beyond) are a few different issues from the fix. 

Completely getting rid of the CurrentlyOpen property is a no go, if having access to it as a private variable is 100% required for a fix to work (though I don't see why it would be) then it would be best to have the property control the private variable, having the private set public get property is important to be able to access it from outside sources. 

Similarly, having OpenMenu only accessible via GameObject isn't good, being able to control them from places that may not have access to the gameobject (modded code being the most likely) is important, though having an overloaded function called with the name that in turn calls the gameObject version would be acceptable.

Also, you seem to have removed passing the prefab name, and hardwired it to use the ConstructionMenu prefab, while those two menus may be similar enough for it to not be a noticeable effect, it is a bug and will lead to more obvious issues in the future, it should go back to passing and using the prefab name as appropriate.

Similarly, you've changed CloseMenu and OpenMunu's check to show/hide the spawn inventory ui and turn on/off building mode, and again this is a case of something that may function similarly for now, those two won't always be the only menus in MenuLeft, and that behavior may not be the desirable behavior for all of them.

Finally, I'm not sure that currentlyOpen only being set from the buttonClick handler is a good idea, that means that if the menus are opened from elsewhere in code, currentlyOpen will no longer be accurate.

onurtnrkl commented:
> Ok, a few issues here. First, animation speed should be set in code rather than prefab.
> 
> Second, it's generally not a good idea to include fixes with feature PRs, unless that fix is integral for the feature to function at all, it means we can't accept the fix until the feature is accepted.
> 
> Third (and beyond) are a few different issues from the fix.
> 
> Completely getting rid of the CurrentlyOpen property is a no go, if having access to it as a private variable is 100% required for a fix to work (though I don't see why it would be) then it would be best to have the property control the private variable, having the private set public get property is important to be able to access it from outside sources.
> 
> Similarly, having OpenMenu only accessible via GameObject isn't good, being able to control them from places that may not have access to the gameobject (modded code being the most likely) is important, though having an overloaded function called with the name that in turn calls the gameObject version would be acceptable.
> 
> Also, you seem to have removed passing the prefab name, and hardwired it to use the ConstructionMenu prefab, while those two menus may be similar enough for it to not be a noticeable effect, it is a bug and will lead to more obvious issues in the future, it should go back to passing and using the prefab name as appropriate.
> 
> Similarly, you've changed CloseMenu and OpenMunu's check to show/hide the spawn inventory ui and turn on/off building mode, and again this is a case of something that may function similarly for now, those two won't always be the only menus in MenuLeft, and that behavior may not be the desirable behavior for all of them.
> 
> Finally, I'm not sure that currentlyOpen only being set from the buttonClick handler is a good idea, that means that if the menus are opened from elsewhere in code, currentlyOpen will no longer be accurate.

Okay, I reverted my fix commit. I am waiting for my PR feature to be accepted.

After that, I will make the changes you asked for my fix commit. 

Thank you.

kd7uiy commented:
Before we can pull it in, you should address the other issues that koosemoose included. Removing the fix was pretty minor compared to the other things. If you would like, feel free to reach out to us more realtime in Discord. https://discord.gg/rmj8PW

koosemose commented:
Actually removing the fix removed most of the issues I had (as the bulk of my comments had to do with things within the fix). 

The only remaining thing is the animation speed being set from the prefab rather than in code. Partially it's just a matter of our standard practices, we tend to prefer in-code solutions where possible. Also, the only reason I can see for having it set from the prefab the script is attached to is if each animation needed to be at a different speed (which would probably feel inconsistent, so not a good thing), and even then having it set in code would be preferable if possible. 

As is, aside from our standard practices, if the optimal visual aesthetic is to have everything move at the same speed, then having to change it in multiple locations would be a pain at best, and at worst could lead to some animations over time ending up with different speeds (from people overlooking one prefab or another).

And as kd7 says, reaching out over our discord can be useful, to work out any minor details or figure out a best approach or whatever (plus as a bonus, I tend to be a lot less wordy in Discord... at least somewhat less)

BraedonWooding commented on Assets/StreamingAssets/CSharp/SettingsMenuFunctions.cs:
```diff
@@ -1093,6 +1093,19 @@ public override void CancelSetting()
+        base.CancelSetting();
+    }
+}
+
```
You don't need to have this :).  Just change the ClassName from `UIAnimationSpeedSlider` to `GenericSlider`

BraedonWooding commented on Assets/StreamingAssets/Localization/en_US.lang:
```diff
@@ -249,6 +250,7 @@ ui_skin_tooltip=Skins for the UI
 quality_tooltip=Quality Presets
 vsync_tooltip=Helps prevent screen tearing by capping frame rate
 ui_scale_tooltip=Scale the UI manually
+ui_animation_speed_tooltip=Addjust the speed of UI animations
```
Typo.  `Adjust`

BraedonWooding commented on Assets/StreamingAssets/Data/Prototypes/SettingsTemplate.json:
```diff
@@ -187,6 +187,17 @@
+            "Name": "ui_animation_speed",
+            "Key": "video_general_ui_animation_speed",
+            "DefaultValue": "5",
+            "ClassName": "UIAnimationSpeedSlider",
```
Since you don't need any custom logic just change it to `GenericSlider`

onurtnrkl replied:
You right, I was thinking about some logic(Apply speed function in controller or something like that) but I will leave it for now. Thank you for reminding that.

BraedonWooding commented:
I'll let @kd7uiy do the final checks and merge it cause he's been following this closer than me and I can't do the final test to check speed of the menu and so on (away from Unity rn).
